### DISCUSSION TITLE
How to set fig.path to include chapter number?

### DISCUSSION THREAD

**Discussion by friendly** [quarto-dev/quarto-cli#7546] [Q&A]
### Description

For a Quarto book project, I'm generating many figures, currently all sent to a single folder, using the knitr `fig.path` option/

```
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  error = FALSE,
  fig.align = 'center',
  fig.path = "figs/"
 ...
```
A file, `common.R` containing this code is sourced at the start of every chapter, via

```{r include=FALSE}
source("R/common.R")
```

I would like to be able to set the `fig.path` on a chapter basis, eg, to `figs/ch01/`, `figs/ch02/` ... I know I can do this by modifying the initial chunk for each chapter to

```{r include=FALSE}
source("R/common.R")
knitr::opts_chunk$set(fig.path = "figs/ch01")
```

But is there any way I can capture the chapter number in my `common.R` script and have it work for all chapters? 

I'm thinking somehow to use `basename()` of the current .qmd file, e.g., `01-intro.qmd` and extract the chapter number from that.
Or else, is there a chapter counter I could access?

**Comment by mcanouil** [comment]
Note that `fig.path` is a knitr option (global if I am not mistaken) which is independent of Quarto.
Also the use of `knitr::opts_chunk$set` is highly discouraged in Quarto as it is rmarkdown/knitr syntax which is not Quarto syntax.
see:
- https://quarto.org/docs/computations/r.html#chunk-options
- https://quarto.org/docs/computations/r.html#knitr-options
- https://quarto.org/docs/computations/execution-options.html

This being said, I don't know a simple way to do what you want.
Possibly using post-processing or using lua filter.

Maybe the future work tracked in #6518 could help with this in the future.

**Reply by cderv** [reply]
> Also the use of knitr::opts_chunk$set is highly discouraged in Quarto as it is rmarkdown/knitr syntax which is not Quarto syntax.

There is really nothing wrong to use `knitr::opts_chunk$set()` in a setup chunk when using Quarto documents with the R engine. 
````markdown
```{r}
#| label: setup
#| include: false 
source("R/common.R")
knitr::opts_chunk$set(fig.path = "figs/ch01")
```
````

The YAML syntax is there mainly to avoid issue when doing multiple language project and to be more consistant with Jupyter and other engines. And internally, using `knitr: opts_chunk:` in YAML will do that exact command. 

When using R only Quarto project, as you do because your are already sourcing a R file in a setup chunk. There is really no problem doing it. So no need to convert it to a Quarto syntax - setting R options, or **knitr** options only through YAML can be challenging sometimes when using R expression as values. 

To be clear, the above is perfectly correct - I just want to tone down a little for R users the reply here regarding "highly discouraged"

**Reply by mcanouil** [reply]
The comment is mostly for "message", "warning", "error", or "fig.align" which has a meaning in the cross-reference system.

**Reply by mcanouil** [reply]
Note that `fig.path` is 100 % specific to knitr which means `knitr::opts_chunk$set(fig.path = "figs/ch01")` is indeed totally fine but `knitr::opts_chunk$set(fig.align = "center")` is not guarantee to not have unintended side effects in Quarto (sorry for the double negation).

**Comment by friendly** [comment]
Thanks for this reply, and for those pointers to the `r.html` docs. The **Note**: Note that if you prefer it is still possible to include chunk options on the first line (e.g. ```{r, echo = FALSE}).  is particularly helpful,

I guess, for my use case, I'll go with manually resetting `fig.path` at the start of each chapter.

**Reply by mcanouil** [reply]
Yes it's possible to set in code cell but it means you bypass Quarto which might not end well.
So you should not rely on this approach when there is a Quarto approach.
Also note that Quarto highly recommended syntax is YAML even for code cell.

````
```{r}
#| echo: false

```
````

For short, if you decide to bypass Quarto settings, then do not blame Quarto if not everything works as you might have expected.

**Comment by baptiste** [comment]
As a really ugly workaround, you could write the current chapter number to a tmp file, and use that value to set the fig path,

````
if(!file.exists('.chapter')) chap <- 1 else
  chap <- as.numeric(readLines('.chapter')[1]) + 1
writeLines(paste(chap), '.chapter')
message(glue::glue("chapter {chap} currently processed"))

knitr::opts_chunk$set(fig.path = glue::glue("figs/chap{chap}/"))
````

(with suitable pre/post-render script to delete the counter file)

Obviously this relies on quarto processing files in the same order as you want, but presumably your naming convention will do that.

**Reply by cscheid** [reply]
> Obviously this relies on quarto processing files in the same order as you want, but presumably your naming convention will do that.

I'll say right now that this will probably stop working in quarto 1.5. We'll add better mechanisms for controlling this kind of numbering, but in quarto 1.5 we plan to add parallel rendering, so "in the same order as you want" will kind of go away as a default option.
